File: python/ack_cost_sdk/allocation.py
"""
Allocation API服务模块
"""

from typing import Optional, Dict, List
from urllib.parse import urlencode
from dataclasses import dataclass, field
import json
from .client import Client
from .models import Properties, CostData, Response
from .errors import CostError, http_error_from_status_code


@dataclass
class AllocationRequest:
    """Allocation API请求参数"""
    window: str                                    # 查询的持续时间
    filter: Optional[str] = None                  # 资源过滤条件
    step: Optional[str] = None                    # 时间分段
    aggregate: Optional[str] = None               # 聚合维度
    idle: Optional[bool] = None                   # 是否展示闲置成本
    share_idle: Optional[bool] = None             # 是否分摊闲置成本
    share_split: Optional[str] = None             # 闲置分摊策略
    idle_by_node: Optional[bool] = None           # 是否按节点维度聚合闲置成本
    target_type: Optional[str] = None             # 成本分摊的目标类型
    format: Optional[str] = None                  # 成本导出格式


@dataclass
class AllocationData(CostData):
    """Allocation 成本数据"""
    pass


@dataclass
class AllocationResponse(Response):
    """Allocation API响应"""
    data: List[Dict[str, AllocationData]] = field(default_factory=list)


class AllocationService:
    """Allocation API服务"""
    
    def __init__(self, client: Client):
        self.client = client
    
    def get_allocation(self, request: AllocationRequest) -> AllocationResponse:
        """
        查询Allocation成本数据
        
        Args:
            request: AllocationRequest请求对象
            
        Returns:
            AllocationResponse: 响应对象
            
        Raises:
            CostError: 请求失败或解析错误
        """
        # 构建URL
        api_url = f"{self.client.config.api_server}/api/v1/namespaces/kube-system/services/ack-metrics-adapter-api-service:8080/proxy/v2/allocation"
        
        # 添加查询参数
        params = {}
        params["window"] = request.window
        
        if request.filter:
            params["filter"] = request.filter
            
        if request.step:
            params["step"] = request.step
            
        if request.aggregate:
            params["aggregate"] = request.aggregate
            
        if request.idle is not None:
            params["idle"] = str(request.idle).lower()
            
        if request.share_idle is not None:
            params["shareIdle"] = str(request.share_idle).lower()
            
        if request.share_split:
            params["shareSplit"] = request.share_split
            
        if request.idle_by_node is not None:
            params["idleByNode"] = str(request.idle_by_node).lower()
            
        if request.target_type:
            params["targetType"] = request.target_type
            
        if request.format:
            params["format"] = request.format
        
        # 构建完整URL
        if params:
            api_url = f"{api_url}?{urlencode(params)}"
        
        try:
            # 发送请求
            response = self.client.do_request("GET", api_url)
            
            # 检查响应状态码
            if response.status_code != 200:
                raise http_error_from_status_code(response.status_code)
            
            # 解析响应
            raw_data = response.json()
            result = AllocationResponse()
            
            # 转换数据格式
            for item in raw_data.get("data", []):
                data_dict = {}
                for key, value in item.items():
                    allocation_data = AllocationData()
                    allocation_data.name = value.get("name", "")
                    
                    # 处理properties
                    props = value.get("properties", {})
                    allocation_data.properties = Properties(
                        pod=props.get("pod"),
                        node=props.get("node"),
                        namespace=props.get("namespace"),
                        controller_kind=props.get("controllerKind"),
                        controller=props.get("controller"),
                        provider_id=props.get("providerID"),
                        labels=props.get("labels", {})
                    )
                    
                    allocation_data.start = value.get("start", "")
                    allocation_data.end = value.get("end", "")
                    allocation_data.cpu_core_request_average = value.get("cpuCoreRequestAverage", 0.0)
                    allocation_data.cpu_core_usage_average = value.get("cpuCoreUsageAverage", 0.0)
                    allocation_data.ram_byte_request_average = value.get("ramByteRequestAverage", 0.0)
                    allocation_data.ram_byte_usage_average = value.get("ramByteUsageAverage", 0.0)
                    allocation_data.cost = value.get("cost", 0.0)
                    allocation_data.cost_ratio = value.get("costRatio", 0.0)
                    allocation_data.custom_cost = value.get("customCost", 0.0)
                    
                    data_dict[key] = allocation_data
                result.data.append(data_dict)
            
            return result
            
        except json.JSONDecodeError as e:
            raise CostError("InternalError", "Failed to decode response", e)
        except Exception as e:
            if isinstance(e, CostError):
                raise e
            raise CostError("InternalError", "Failed to send request", e)

File: go/allocation.go
package cost

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"strconv"
)

// AllocationService Allocation API服务
type AllocationService struct {
	client *Client
}

// AllocationProperties Kubernetes对象属性
type AllocationProperties struct {
	Pod            string            `json:"pod,omitempty"`
	Node           string            `json:"node,omitempty"`
	Namespace      string            `json:"namespace,omitempty"`
	ControllerKind string            `json:"controllerKind,omitempty"`
	Controller     string            `json:"controller,omitempty"`
	ProviderID     string            `json:"providerID,omitempty"`
	Labels         map[string]string `json:"labels,omitempty"`
}

// AllocationData 分摊成本数据
type AllocationData struct {
	Name                 string              `json:"name"`
	Properties           AllocationProperties `json:"properties"`
	Start                string               `json:"start"`
	End                  string               `json:"end"`
	CPUCoreRequestAverage float64            `json:"cpuCoreRequestAverage"`
	CPUCoreUsageAverage  float64             `json:"cpuCoreUsageAverage"`
	RAMByteRequestAverage float64           `json:"ramByteRequestAverage"`
	RAMByteUsageAverage  float64             `json:"ramByteUsageAverage"`
	Cost                 float64             `json:"cost"`
	CostRatio            float64             `json:"costRatio"`
	CustomCost           float64             `json:"customCost"`
}

// AllocationResponse Allocation API响应
type AllocationResponse struct {
	Data []map[string]*AllocationData `json:"data"`
}

// AllocationRequest Allocation API请求参数
type AllocationRequest struct {
	// Window 查询的持续时间
	Window string `url:"window"`
	
	// Filter 资源过滤条件
	Filter string `url:"filter,omitempty"`
	
	// Step 时间分段
	Step string `url:"step,omitempty"`
	
	// Aggregate 聚合维度
	Aggregate string `url:"aggregate,omitempty"`
	
	// Idle 是否展示闲置成本
	Idle *bool `url:"idle,omitempty"`
	
	// ShareIdle 是否分摊闲置成本
	ShareIdle *bool `url:"shareIdle,omitempty"`
	
	// ShareSplit 闲置分摊策略
	ShareSplit string `url:"shareSplit,omitempty"`
	
	// IdleByNode 是否按节点维度聚合闲置成本
	IdleByNode *bool `url:"idleByNode,omitempty"`
	
	// TargetType 成本分摊的目标类型
	TargetType string `url:"targetType,omitempty"`
	
	// Format 成本导出格式
	Format string `url:"format,omitempty"`
}

// GetAllocation 查询Allocation成本数据
func (s *AllocationService) GetAllocation(ctx context.Context, req *AllocationRequest) (*AllocationResponse, error) {
	// 构建URL
	apiURL := fmt.Sprintf("%s/api/v1/namespaces/kube-system/services/ack-metrics-adapter-api-service:8080/proxy/v2/allocation", s.client.config.APIServer)
	
	// 添加查询参数
	params := url.Values{}
	params.Add("window", req.Window)
	
	if req.Filter != "" {
		params.Add("filter", req.Filter)
	}
	
	if req.Step != "" {
		params.Add("step", req.Step)
	}
	
	if req.Aggregate != "" {
		params.Add("aggregate", req.Aggregate)
	}
	
	if req.Idle != nil {
		params.Add("idle", strconv.FormatBool(*req.Idle))
	}
	
	if req.ShareIdle != nil {
		params.Add("shareIdle", strconv.FormatBool(*req.ShareIdle))
	}
	
	if req.ShareSplit != "" {
		params.Add("shareSplit", req.ShareSplit)
	}
	
	if req.IdleByNode != nil {
		params.Add("idleByNode", strconv.FormatBool(*req.IdleByNode))
	}
	
	if req.TargetType != "" {
		params.Add("targetType", req.TargetType)
	}
	
	if req.Format != "" {
		params.Add("format", req.Format)
	}
	
	// 构建完整URL
	if len(params) > 0 {
		apiURL = fmt.Sprintf("%s?%s", apiURL, params.Encode())
	}
	
	// 创建HTTP请求
	httpReq, err := http.NewRequest("GET", apiURL, nil)
	if err != nil {
		return nil, NewError(ErrCodeBadRequest, "Failed to create request", err)
	}
	
	// 发送请求
	httpResp, err := s.client.Do(ctx, httpReq)
	if err != nil {
		return nil, NewError(ErrCodeInternalError, "Failed to send request", err)
	}
	defer httpResp.Body.Close()
	
	// 检查响应状态码
	if httpResp.StatusCode != http.StatusOK {
		return nil, HTTPErrorFromStatusCode(httpResp.StatusCode, nil)
	}
	
	// 解析响应
	var resp AllocationResponse
	if err := json.NewDecoder(httpResp.Body).Decode(&resp); err != nil {
		return nil, NewError(ErrCodeInternalError, "Failed to decode response", err)
	}
	
	return &resp, nil
}